LandMark/Feature/Persistence/CoreDataDataSource.swift

```diff
@@ -0,0 +1,63 @@
+//
+//  CoreDataDataSource.swift
+//  LandMark
+//
+//  Created by Lucas Migge on 31/10/24.
+//
+
+import CoreData
+
+protocol CoreDataDataSourceProtocol {
+    associatedtype Entity = NSManagedObject
+    func create(_ entity: Entity)
+    func fetchAll() -> [Entity]
+    func fetch(byID id: NSManagedObjectID) -> Entity?
+    func update(_ entity: Entity)
+    func delete(_ entity: Entity)
+}
+
+class CoreDataDataSource<T: NSManagedObject>: CoreDataDataSourceProtocol {
+    let context: NSManagedObjectContext
+    
+    init(context: NSManagedObjectContext = PersistenceController.shared.container.viewContext) {
+        self.context = context
+    }
+    
+    func create(_ entity: T) {
+        context.insert(entity)
+        saveContext()
+    }
+    
+    func fetchAll() -> [T] {
+        let request = T.fetchRequest()
+        do {
+            return try context.fetch(request) as? [T] ?? []
+        } catch {
+            print("Error fetching data: \(error)")
+            return []
+        }
+    }
```

Comment:
Mesma coisa sobre propagar os possíveis erros.

```suggestion
    func fetchAll() -> [T] {
        let request = T.fetchRequest()
        do {
            // Aqui acho que também vale um erro pra esse caso, de não conseguir fazer o cast, also como:
            guard let entities = try context.fetch(request) as? [T] else {
                return .failure(CoreDataDataSourceError.couldNotCastType)
            }
            
            return .success(entities)
        } catch {
            return .failure(CoreDataDataSourceError.failureFetchingData)
        }
    }
```